pkg/querier/frontend/frontend.go

```diff
@@ -32,6 +32,7 @@ import (
 const (
 	// StatusClientClosedRequest is the status code for when a client request cancellation of an http request
 	StatusClientClosedRequest = 499
+	maxBodySize               = 10 * 1024 * 1024 // 10 MiB
```

Comment:
**thought(non-blocking):** This directly relates to an issue that Prometheus/Cortex users currently experience when they have a datasource configured to use a `GET` method. There is a `1 MiB` limit configured on the MaxHeaderSize in the Go server and queries with a large number of regex matchers will hit this limit causing queries to fail. The workaround is to switch the datasource to use a `POST` method. However, after this change, if a user hits this limit they will not have a readily available workaround. I personally think 10 MiB is reasonable as a max.

**suggestion:** I think a comment here and/or in the changelog denoting that this potentially could cause failures for large query strings with a very large number of regex matchers.

Reply:
That made me think that this change is really dangerous as it will fail silently and potential users will be puzzled why some things doesn't work or give unexpected results. 

I think we could:

- add warning if Content-Length header says it's larger than 10MiB - without reading the body before the RoundTrip
- remove the limit for now
- leave it as is with 10MiB limit (no idea how many people run into request's body larger than 10 MiB 🤔 )

Reply:
> That made me think that this change is really dangerous as it will fail silently and potential users will be puzzled why some things doesn't work or give unexpected results.

It doesn't fail silently, but with "http: request body too large" error, however we would send it back with status code 500, which isn't so good, since it's a client error and we should use 413 instead.

Unfortunately relevant error isn't public, so we could either 1) check for the error string from stdlib, which would be an ugly hack but it will work and with a good test we will catch change of the string easily, or 2) rework this completely, but nice thing about using `http.MaxBytesReader` is that it will also close the connection and not try to read the rest of the request. Personally I would be fine with approach 1) in `writeError` method (incl. the test for error reported by `http.MaxBytesReader`).

> I think we could:
> 
> * add warning if Content-Length header says it's larger than 10MiB - without reading the body before the RoundTrip

That's not enough, chunked requests have no content-length.

> * remove the limit for now

Query-frontend has no limit right now, and sending large enough request *does* crash it (easy to check using curl).

> * leave it as is with 10MiB limit (no idea how many people run into request's body larger than 10 MiB 🤔 )

I did a check on our clusters, and we don't receive query requests larger than 1 MiB. I suggested to use config option just in case, but I haven't insisted on using it since it never happens.

Reply:
> It doesn't fail silently, but with "http: request body too large" error, however we would send it back with status code 500, which isn't so good, since it's a client error and we should use 413 instead.

Yes, you are absolutely right, had LimitReader docs in head while writing this 😆 

I will add `"http: request body too large"` error handling in `writeError` and write a test for this.

Reply:
I have added new option to control body size and also handling for `http: request body too large`.

Currently the error message will look like this: `http: request body too large` but the original message is `readfrom tcp 127.0.0.1:54370->127.0.0.1:54365: http: request body too large`, my question is do we leave it as is or we create error inline with original message.